velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "HdfsReadFile.h"
+#include <folly/synchronization/CallOnce.h>
+#include <hdfs/hdfs.h>
+
+namespace facebook::velox {
+
+HdfsReadFile::HdfsReadFile(hdfsFS hdfs, std::string_view path)
+    : hdfsClient_(hdfs), filePath_(path) {
+  fileInfo_ = hdfsGetPathInfo(hdfsClient_, filePath_.data());
+  VELOX_CHECK_NOT_NULL(
+      fileInfo_,
+      "Unable to get file path info for file: {}. got error: {}",
+      filePath_,
+      hdfsGetLastError());
+}
+
+void HdfsReadFile::preadInternal(uint64_t offset, uint64_t length, char* pos)
+    const {
+  checkFileReadParameters(offset, length);
+  auto file = hdfsOpenFile(hdfsClient_, filePath_.data(), O_RDONLY, 0, 0, 0);
```

Comment:
Opening the file for every `read` is not lightweight correct?. If I recall, the library has to talk to get NameNode and get a list of block locations which then are used to open the file. We need to figure out a way to avoid this.

Reply:
`hdfsOpenFile` returns a struct with a input/output stream. 

The block location logic actually happens on `hdfsRead` 

Inside `hdfsRead` there is logic to locate blocks:

1)Get block information from namenode
2)Check if we reach the end of file
3)seek to the right block to read
4) setup block reader